File: BDFM-RAG-System/src/services/qdrant.service.ts
import { QdrantClient } from '@qdrant/js-client-rest';
import config from '../config';
import logger from '../utils/logger';
import { SearchResult, VectorEmbedding, SearchFilters } from '../models';

export class QdrantService {
  private client: QdrantClient;
  private initialized: boolean = false;

  constructor() {
    this.client = new QdrantClient({
      url: config.qdrant.url,
      apiKey: config.qdrant.apiKey,
    });
  }

  /**
   * Initialize Qdrant and create collections
   */
  async initialize(): Promise<void> {
    try {
      logger.info('Initializing Qdrant service...');

      // Check if collections exist, create if not
      await this.ensureCollection(
        config.collections.correspondence,
        config.rag.embeddingDimension
      );
      await this.ensureCollection(
        config.collections.workflow,
        config.rag.embeddingDimension
      );
      await this.ensureCollection(
        config.collections.userGuide,
        config.rag.embeddingDimension
      );

      this.initialized = true;
      logger.info('Qdrant service initialized successfully');
    } catch (error) {
      logger.error('Failed to initialize Qdrant service:', error);
      throw error;
    }
  }

  /**
   * Ensure collection exists
   */
  private async ensureCollection(
    collectionName: string,
    vectorSize: number
  ): Promise<void> {
    try {
      const collections = await this.client.getCollections();
      const exists = collections.collections.some(
        (col) => col.name === collectionName
      );

      if (!exists) {
        logger.info(`Creating collection: ${collectionName}`);
        await this.client.createCollection(collectionName, {
          vectors: {
            size: vectorSize,
            distance: 'Cosine',
          },
          optimizers_config: {
            default_segment_number: 2,
          },
          replication_factor: 1,
        });
        logger.info(`Collection created: ${collectionName}`);
      } else {
        logger.info(`Collection already exists: ${collectionName}`);
      }
    } catch (error) {
      logger.error(`Error ensuring collection ${collectionName}:`, error);
      throw error;
    }
  }

  /**
   * Upsert embeddings to collection
   */
  async upsertEmbeddings(
    collectionName: string,
    embeddings: VectorEmbedding[]
  ): Promise<void> {
    try {
      const points = embeddings.map((emb) => ({
        id: emb.id,
        vector: emb.embeddingVector,
        payload: {
          correspondenceId: emb.correspondenceId,
          textChunk: emb.textChunk,
          modelName: emb.modelName,
          chunkIndex: emb.chunkIndex,
          language: emb.language,
          createdAt: emb.createdAt.toISOString(),
          ...emb.metadata,
        },
      }));

      await this.client.upsert(collectionName, {
        wait: true,
        points,
      });

      logger.info(
        `Upserted ${embeddings.length} embeddings to ${collectionName}`
      );
    } catch (error) {
      logger.error(`Error upserting embeddings to ${collectionName}:`, error);
      throw error;
    }
  }

  /**
   * Search similar vectors
   */
  async searchSimilar(
    collectionName: string,
    queryVector: number[],
    limit: number = 10,
    threshold: number = 0.7,
    filters?: SearchFilters
  ): Promise<SearchResult[]> {
    try {
      const filter = this.buildFilter(filters);

      const searchResult = await this.client.search(collectionName, {
        vector: queryVector,
        limit,
        score_threshold: threshold,
        with_payload: true,
        filter,
      });

      const results: SearchResult[] = searchResult.map((result) => ({
        id: result.payload?.correspondenceId as string,
        mailNum: result.payload?.mail_num as string,
        mailDate: result.payload?.mail_date as string,
        subject: result.payload?.subject as string,
        bodyText: result.payload?.body_text as string,
        correspondenceType: result.payload?.correspondence_type as string,
        secrecyLevel: result.payload?.secrecy_level as string,
        priorityLevel: result.payload?.priority_level as string,
        personalityLevel: result.payload?.personality_level as string,
        similarityScore: result.score,
      }));

      logger.info(
        `Found ${results.length} similar results in ${collectionName}`
      );
      return results;
    } catch (error) {
      logger.error(`Error searching in ${collectionName}:`, error);
      throw error;
    }
  }

  /**
   * Delete embeddings by correspondence ID
   */
  async deleteByCorrespondenceId(
    collectionName: string,
    correspondenceId: string
  ): Promise<void> {
    try {
      await this.client.delete(collectionName, {
        wait: true,
        filter: {
          must: [
            {
              key: 'correspondenceId',
              match: {
                value: correspondenceId,
              },
            },
          ],
        },
      });

      logger.info(
        `Deleted embeddings for correspondence ${correspondenceId} from ${collectionName}`
      );
    } catch (error) {
      logger.error(
        `Error deleting embeddings for ${correspondenceId}:`,
        error
      );
      throw error;
    }
  }

  /**
   * Get collection info
   */
  async getCollectionInfo(collectionName: string): Promise<any> {
    try {
      return await this.client.getCollection(collectionName);
    } catch (error) {
      logger.error(`Error getting collection info for ${collectionName}:`, error);
      throw error;
    }
  }

  /**
   * Get collection count
   */
  async getCollectionCount(collectionName: string): Promise<number> {
    try {
      const info = await this.getCollectionInfo(collectionName);
      return info.points_count || 0;
    } catch (error) {
      logger.error(`Error getting collection count for ${collectionName}:`, error);
      return 0;
    }
  }

  /**
   * Delete collection
   */
  async deleteCollection(collectionName: string): Promise<void> {
    try {
      await this.client.deleteCollection(collectionName);
      logger.info(`Deleted collection: ${collectionName}`);
    } catch (error) {
      logger.error(`Error deleting collection ${collectionName}:`, error);
      throw error;
    }
  }

  /**
   * Build filter for search
   */
  private buildFilter(filters?: SearchFilters): any {
    if (!filters) return undefined;

    const must: any[] = [];

    if (filters.correspondenceType && filters.correspondenceType.length > 0) {
      must.push({
        key: 'correspondence_type',
        match: {
          any: filters.correspondenceType,
        },
      });
    }

    if (filters.priorityLevel && filters.priorityLevel.length > 0) {
      must.push({
        key: 'priority_level',
        match: {
          any: filters.priorityLevel,
        },
      });
    }

    if (filters.secrecyLevel && filters.secrecyLevel.length > 0) {
      must.push({
        key: 'secrecy_level',
        match: {
          any: filters.secrecyLevel,
        },
      });
    }

    if (filters.dateFrom) {
      must.push({
        key: 'mail_date',
        range: {
          gte: filters.dateFrom,
        },
      });
    }

    if (filters.dateTo) {
      must.push({
        key: 'mail_date',
        range: {
          lte: filters.dateTo,
        },
      });
    }

    if (filters.organizationalUnitId) {
      must.push({
        key: 'organizational_unit_id',
        match: {
          value: filters.organizationalUnitId,
        },
      });
    }

    return must.length > 0 ? { must } : undefined;
  }

  /**
   * Check connection
   */
  async checkConnection(): Promise<boolean> {
    try {
      await this.client.getCollections();
      return true;
    } catch (error) {
      logger.error('Qdrant connection check failed:', error);
      return false;
    }
  }

  /**
   * Get all collection names
   */
  async getCollections(): Promise<string[]> {
    try {
      const collections = await this.client.getCollections();
      return collections.collections.map((col) => col.name);
    } catch (error) {
      logger.error('Error getting collections:', error);
      return [];
    }
  }
}

// Export singleton instance
export default new QdrantService();

File: BDFM-RAG-System/bdfm-rag-python/app/services/qdrant_service.py
"""Qdrant vector database service"""
from qdrant_client import QdrantClient
from qdrant_client.http.exceptions import UnexpectedResponse
from qdrant_client.models import (
    Distance,
    VectorParams,
    PointStruct,
    Filter,
    FieldCondition,
    MatchValue,
    Range,
)
from typing import List, Dict, Any, Optional
from app.config import settings
from app.utils.logger import logger


class QdrantService:
    """Service for interacting with Qdrant vector database"""
    
    def __init__(self):
        self.client: Optional[QdrantClient] = None
        self.dimension = settings.embedding_dimension
        
    def initialize(self):
        """Initialize Qdrant client and create collections"""
        try:
            logger.info(f"Connecting to Qdrant at {settings.qdrant_url}")
            
            self.client = QdrantClient(
                url=settings.qdrant_url,
                api_key=settings.qdrant_api_key,
                timeout=60,
                # Older Qdrant servers may be incompatible with newer clients; skip strict check
                check_compatibility=False,
                prefer_grpc=False,
            )
            
            logger.info("Qdrant client initialized successfully")
            
            # Create collections if they don't exist
            self._create_collections()
            
        except Exception as e:
            logger.error(f"Error initializing Qdrant: {e}")
            raise
    
    def _create_collections(self):
        """Create Qdrant collections if they don't exist"""
        collections = [
            settings.correspondence_collection,
            settings.workflow_collection,
            settings.user_guide_collection,
        ]
        
        for collection_name in collections:
            try:
                # Check if collection exists (compat with older Qdrant servers)
                if not self._collection_exists_compat(collection_name):
                    logger.info(f"Creating collection: {collection_name}")
                    
                    self.client.create_collection(
                        collection_name=collection_name,
                        vectors_config=VectorParams(
                            size=self.dimension,
                            distance=Distance.COSINE,
                        ),
                    )
                    
                    logger.info(f"Collection {collection_name} created successfully")
                else:
                    logger.info(f"Collection {collection_name} already exists")
                    
            except Exception as e:
                logger.error(f"Error creating collection {collection_name}: {e}")
                raise

    def _collection_exists_compat(self, collection_name: str) -> bool:
        """Check if a collection exists, compatible with older Qdrant servers.

        Some older server versions do not support the dedicated
        collection_exists endpoint and will return 404 for it. In that case,
        attempt to fetch collection info and treat 404 as non-existence.
        """
        try:
            # Preferred quick path where supported
            if hasattr(self.client, "collection_exists"):
                try:
                    return bool(self.client.collection_exists(collection_name))
                except UnexpectedResponse as ex:
                    # Fall back on 404 from older server implementations
                    # UnexpectedResponse doesn't always expose status code; rely on message
                    message = str(ex).lower()
                    if "404" in message or "not found" in message:
                        return False
                    raise

            # Fallback: try get_collection
            self.client.get_collection(collection_name)
            return True
        except UnexpectedResponse as ex:
            message = str(ex).lower()
            if "404" in message or "not found" in message:
                return False
            raise
        except Exception as ex:
            message = str(ex).lower()
            if "404" in message or "not found" in message:
                return False
            # Unknown error should propagate
            raise
    
    async def upsert_embeddings(
        self,
        collection_name: str,
        embeddings: List[Dict[str, Any]]
    ) -> bool:
        """
        إضافة أو تحديث embeddings في Qdrant
        
        Args:
            collection_name: اسم الـ collection
            embeddings: قائمة embeddings
            
        Returns:
            True if successful
        """
        try:
            if not embeddings:
                return True
            
            points = []
            for emb in embeddings:
                point = PointStruct(
                    id=emb['id'],
                    vector=emb['embedding_vector'],
                    payload={
                        'correspondence_id': emb['correspondence_id'],
                        'text_chunk': emb['text_chunk'],
                        'chunk_index': emb['chunk_index'],
                        **emb['metadata']
                    }
                )
                points.append(point)
            
            self.client.upsert(
                collection_name=collection_name,
                points=points
            )
            
            logger.debug(f"Upserted {len(points)} points to {collection_name}")
            return True
            
        except Exception as e:
            logger.error(f"Error upserting embeddings: {e}")
            raise
    
    async def search_similar(
        self,
        collection_name: str,
        query_vector: List[float],
        limit: int = 10,
        score_threshold: float = 0.7,
        filters: Optional[Dict[str, Any]] = None
    ) -> List[Dict[str, Any]]:
        """
        البحث عن embeddings مشابهة
        
        Args:
            collection_name: اسم الـ collection
            query_vector: vector الاستعلام
            limit: عدد النتائج
            score_threshold: عتبة التشابه
            filters: فلاتر إضافية
            
        Returns:
            قائمة النتائج
        """
        try:
            # Build filter if provided
            query_filter = None
            if filters:
                query_filter = self._build_filter(filters)
            
            # Search
            results = self.client.search(
                collection_name=collection_name,
                query_vector=query_vector,
                limit=limit,
                score_threshold=score_threshold,
                query_filter=query_filter,
            )
            
            # Convert to dict format
            search_results = []
            for result in results:
                search_results.append({
                    'id': result.payload.get('correspondence_id'),
                    'mail_num': result.payload.get('mail_num'),
                    'mail_date': result.payload.get('mail_date'),
                    'subject': result.payload.get('subject'),
                    'body_text': result.payload.get('text_chunk', ''),
                    'correspondence_type': result.payload.get('correspondence_type'),
                    'secrecy_level': result.payload.get('secrecy_level'),
                    'priority_level': result.payload.get('priority_level'),
                    'personality_level': result.payload.get('personality_level'),
                    'similarity_score': result.score,
                })
            
            logger.debug(f"Found {len(search_results)} results in {collection_name}")
            return search_results
            
        except Exception as e:
            logger.error(f"Error searching in Qdrant: {e}")
            raise
    
    def _build_filter(self, filters: Dict[str, Any]) -> Filter:
        """بناء Qdrant filter من dictionary"""
        conditions = []
        
        if filters.get('correspondence_type'):
            conditions.append(
                FieldCondition(
                    key="correspondence_type",
                    match=MatchValue(value=filters['correspondence_type'])
                )
            )
        
        if filters.get('priority_level'):
            conditions.append(
                FieldCondition(
                    key="priority_level",
                    match=MatchValue(value=filters['priority_level'])
                )
            )
        
        # يمكن إضافة المزيد من الشروط هنا
        
        if conditions:
            return Filter(must=conditions)
        
        return None
    
    async def delete_by_correspondence_id(
        self,
        collection_name: str,
        correspondence_id: str
    ) -> bool:
        """
        حذف embeddings لمراسلة معينة
        
        Args:
            collection_name: اسم الـ collection
            correspondence_id: معرف المراسلة
            
        Returns:
            True if successful
        """
        try:
            self.client.delete(
                collection_name=collection_name,
                points_selector=Filter(
                    must=[
                        FieldCondition(
                            key="correspondence_id",
                            match=MatchValue(value=correspondence_id)
                        )
                    ]
                )
            )
            
            logger.debug(f"Deleted embeddings for correspondence {correspondence_id}")
            return True
            
        except Exception as e:
            logger.error(f"Error deleting from Qdrant: {e}")
            raise
    
    async def get_collection_count(self, collection_name: str) -> int:
        """Get count of points in collection"""
        try:
            collection_info = self.client.get_collection(collection_name)
            return collection_info.points_count
        except Exception as e:
            logger.error(f"Error getting collection count: {e}")
            return 0
    
    async def get_collection_info(self, collection_name: str) -> Dict[str, Any]:
        """Get collection information"""
        try:
            collection_info = self.client.get_collection(collection_name)
            return {
                'name': collection_name,
                'points_count': collection_info.points_count,
                'vector_size': collection_info.config.params.vectors.size,
                'distance': collection_info.config.params.vectors.distance,
                'status': collection_info.status,
            }
        except Exception as e:
            logger.error(f"Error getting collection info: {e}")
            raise
    
    async def get_collections(self) -> List[str]:
        """Get list of all collections"""
        try:
            collections = self.client.get_collections()
            return [col.name for col in collections.collections]
        except Exception as e:
            logger.error(f"Error getting collections: {e}")
            return []
    
    async def delete_collection(self, collection_name: str) -> bool:
        """Delete a collection"""
        try:
            self.client.delete_collection(collection_name)
            logger.info(f"Deleted collection: {collection_name}")
            return True
        except Exception as e:
            logger.error(f"Error deleting collection: {e}")
            return False
    
    async def check_connection(self) -> bool:
        """Check Qdrant connection"""
        try:
            collections = self.client.get_collections()
            return True
        except Exception as e:
            logger.error(f"Qdrant connection check failed: {e}")
            return False


# Create global qdrant service instance
qdrant_service = QdrantService()


__all__ = ["qdrant_service", "QdrantService"]

